**Discretización**

**Edad**

In [ ]:
x     y
25    1 (reclaman)
35    0
41    0
28    1 (reclaman)
29    1 (reclaman)
.
.
.
10 k casos

**Naive Bayes**: basado en la **probabilidad condicional**

In [ ]:
Pr(x=¿?|y=1)

In [ ]:
Pr(x=20|y=1)=0.1
Pr(x=21|y=1)=0.01
Pr(x=22|y=1)=0.1
.
.
.
Pr(x=58|y=1)=0.005

**¿Cómo discretizo una variable?**

In [ ]:
1 = 20-30
2 = 31-45
3 = +45

In [ ]:
Pr(x=1|y=1)=0.25
Pr(x=2|y=1)=0.50
Pr(x=3|y=1)=0.25

**Árboles de decisión**

**impureza**: error del árbol (entropia o gini)

In [ ]:
a=6
d=4

impureza inicial

**Nota**: las variables que se agregan al modelo de árbol son discretas, si no lo son el modelo de árbol las discretiza utilizando entropia.

**Discretización por intervalos de igual amplitud**

In [29]:
from sklearn.preprocessing import KBinsDiscretizer
import math as m
import pandas as pd
import numpy as np

In [30]:
data = pd.read_csv("D:/Python/1. Nivel I/7/datos/data.csv")

In [31]:
data.head()

Unnamed: 0      rsrp  redireccion       cqi  incidentes
0           0 -116.5000  90890.42778  8.138370           0
1           1 -115.1875  91193.46144  9.140266           0
2           2 -119.1250  91877.50918  7.808917           0
3           3 -111.8125  92049.63360  8.878271           0
4           4 -115.1875  92275.58028  9.051622           0

In [32]:
data.drop(columns="Unnamed: 0",axis=1, inplace=True)

In [33]:
data.head()

rsrp  redireccion       cqi  incidentes
0 -116.5000  90890.42778  8.138370           0
1 -115.1875  91193.46144  9.140266           0
2 -119.1250  91877.50918  7.808917           0
3 -111.8125  92049.63360  8.878271           0
4 -115.1875  92275.58028  9.051622           0

In [9]:
data["cqi"].min()

4.0703296661377

In [10]:
data["cqi"].max()

11.62320369482

In [11]:
rango = data["cqi"].max() - data["cqi"].min(); print(rango)

7.552874028682301


In [16]:
(4*(rango/4)) + 4.0703296661377

11.62320369482

In [17]:
rango/4

1.8882185071705753

In [ ]:
4.0703296661377 - 5.958548173308275  ---> categoría 1
5.958548173308275 - 7.846766680478851 --> categoría 2
7.846766680478851 - 9.734985187649425 --> categoría 3
9.734985187649425 - 11.62320369482 -->    categoría 4

**¿Cómo determino el número de categorías?**

In [34]:
sturges = round(1+ m.log2(len(data)),0); print(sturges)

12.0


In [19]:
rango/sturges

0.6294061690568584

**función KBinsDiscretizer()**

**Nota**: este proceso de discretización está afectado por la presencia de valores atípicos. No es recomendable aplicarlo cuando ustedes noten la presencia de valores atípicos

In [35]:
amplitud =  KBinsDiscretizer(n_bins=12, #número de categorías de la variable
                            encode="ordinal", #son de naturaleza ordinal
                            strategy="uniform") #uniform= discretización por intervalos de igual amplitud

In [36]:
data["cqi_inter_amplitud"]= amplitud.fit_transform(data[["cqi"]])

In [37]:
data.head()

rsrp  redireccion       cqi  incidentes  cqi_inter_amplitud
0 -116.5000  90890.42778  8.138370           0                 6.0
1 -115.1875  91193.46144  9.140266           0                 8.0
2 -119.1250  91877.50918  7.808917           0                 5.0
3 -111.8125  92049.63360  8.878271           0                 7.0
4 -115.1875  92275.58028  9.051622           0                 7.0

**Discretización por intervalos de igual frecuencia o discretización cuantil**

In [23]:
datan = [4,3,2,6,8,2,5,6]

**Paso 1**: de forma ascendente los valores de la variable

In [24]:
datan = [2,2,3,4,5,6,6,8]

**Paso 2: cuartiles [0.25,0.50,0.75,1]**

In [ ]:
categoría 1 = [2,2]
categoría 2 = [3,4]
categoría 3 = [5,6,6]
categoría 4 = [8]

In [38]:
cuartil =  KBinsDiscretizer(n_bins=4, #número de categorías de la variable
                            encode="ordinal", #son de naturaleza ordinal
                            strategy="quantile") #quantile= discretización por intervalos de igual frecuencia

In [39]:
data["cqi_cuartil"]= cuartil.fit_transform(data[["cqi"]])

In [40]:
data.head()

rsrp  redireccion       cqi  incidentes  cqi_inter_amplitud  \
0 -116.5000  90890.42778  8.138370           0                 6.0   
1 -115.1875  91193.46144  9.140266           0                 8.0   
2 -119.1250  91877.50918  7.808917           0                 5.0   
3 -111.8125  92049.63360  8.878271           0                 7.0   
4 -115.1875  92275.58028  9.051622           0                 7.0   

   cqi_cuartil  
0          1.0  
1          2.0  
2          0.0  
3          2.0  
4          2.0

**Discretización por KMeans**

<center><img src="https://miro.medium.com/max/963/1*Qu9vtRFFrQ1_4jj4Nu__BQ.jpeg" width="700"></center>

In [41]:
datos = [23,25,22,29,36,41,38,18]

In [42]:
c1= 26
c2= 32

In [43]:
data_cl = [1,1,1,2,2,2,2,1]

In [ ]:
c1= 22
c2= 36

In [45]:
data_cl = [1,1,1,1,2,2,2,1]

In [ ]:
c1= 23.4
c2= 38.33

In [48]:
data_cl = [1,1,1,1,2,2,2,1]

**El algoritmo se detiene porque las observaciones dejan de reasignarse a otro cluster**

In [52]:
df = pd.DataFrame(datos,columns=["edad"])

In [53]:
df

edad
0    23
1    25
2    22
3    29
4    36
5    41
6    38
7    18

In [54]:
kmeans =  KBinsDiscretizer(n_bins=2, #número de categorías de la variable
                            encode="ordinal", #son de naturaleza ordinal
                            strategy="kmeans") #kmeans= discretización por k-medias

In [57]:
df["edad_kmeans"]=kmeans.fit_transform(df[["edad"]])+1

In [58]:
df

edad  edad_kmeans
0    23          1.0
1    25          1.0
2    22          1.0
3    29          1.0
4    36          2.0
5    41          2.0
6    38          2.0
7    18          1.0

In [ ]:
data_cl = [1,1,1,1,2,2,2,1]

**Nota**: K-means está afecta a la presencia de valores atípicos

In [59]:
kmeans =  KBinsDiscretizer(n_bins=2, #número de categorías de la variable
                            encode="ordinal", #son de naturaleza ordinal
                            strategy="kmeans") #kmeans= discretización por k-medias

In [60]:
data["cqi_kmeans"]= kmeans.fit_transform(data[["cqi"]])

In [61]:
data.head()

rsrp  redireccion       cqi  incidentes  cqi_inter_amplitud  \
0 -116.5000  90890.42778  8.138370           0                 6.0   
1 -115.1875  91193.46144  9.140266           0                 8.0   
2 -119.1250  91877.50918  7.808917           0                 5.0   
3 -111.8125  92049.63360  8.878271           0                 7.0   
4 -115.1875  92275.58028  9.051622           0                 7.0   

   cqi_cuartil  cqi_kmeans  
0          1.0         0.0  
1          2.0         1.0  
2          0.0         0.0  
3          2.0         1.0  
4          2.0         1.0

**Discretización por entropia o discretización basada en árboles**

- basada en un modelo de aprendizaje : árboles de decisión
- la discretización de la variable x se realizará en función a la variable y  (target)

In [ ]:
x     y
25    1 (reclaman)
35    0
41    0
28    1 (reclaman)
29    1 (reclaman)
.
.
.
10 k casos


<center><img src="https://encrypted-tbn0.gstatic.com/images?q=tbn:ANd9GcTOGIWR9duZ9cM0Rv8mi6WaN0bEpi11oI81Gg&usqp=CAU" width="400"></center>

**Paso 1**: calcular la impureza inicial del árbol por entropia

In [62]:
r=4000
nr=6000
t=10000

In [67]:
impureza=abs(((r/t)*m.log(r/t))+((nr/t)*m.log(nr/t)));print(impureza)

0.6730116670092565


**Se va discretizar la variable x utilizando diferentes puntos de corte (puntos de discretización) y ganará el que reduzca la impureza del árbol**

La impureza inicial del árbol es 0.673 y la impureza final usando el punto de corte "30 años" es 0.672

La impureza inicial del árbol es 0.673 y la impureza final usando el punto de corte "38 años" es 0.35

La impureza inicial del árbol es 0.673 y la impureza final usando el punto de corte "40 años" es 0.27

**Paso 3** seleccionamos el punto de corte de discretización

**Nota** se selecciona el punto de corte que reduzca la impureza inicial del árbol de forma más significativa

**Nodo 1** de 18 a 30 años

In [77]:
t=3500
r=1500
nr=2000

In [78]:
impureza1=abs(((r/t)*m.log(r/t))+((nr/t)*m.log(nr/t)));print(impureza1)

0.6829081047004717


In [79]:
impureza1*0.35

0.23901783664516507

**Nodo 2** de 31 a más años

In [73]:
t=6500
r=2500
nr=4000

In [75]:
impureza2=abs(((r/t)*m.log(r/t))+((nr/t)*m.log(nr/t)));print(impureza2)

0.666278442414676


In [76]:
impureza2*0.65

0.43308098756953944

**calculo de la impureza final de árbol**

In [80]:
suma_i=(impureza1*0.35)+(impureza2*0.65);print(suma_i)

0.6720988242147046


In [1]:
import os 
import pandas as pd 
from sklearn.model_selection import train_test_split

In [ ]:
os.chdir(R "C:\Users\sergi\OneDrive\Documentos\SIR_personal\Python_SDC_HP-i3\data_credit.csv")